R/calculate_recommended_interventions.R

```diff
@@ -302,26 +372,275 @@ calculate_recommended_interventions <- function(
+  if (include_center_effects && !is.null(center_characteristics)) {
+    stop(paste(
+      "Fixed center effects and center characteristics cannot be",
+      "both included in the outcome model. Please",
+      "pick one to use."
+    ))
+  }
+
+  # check if the input data has the "center" column if
+  # include_center_effects is TRUE
+  if (include_center_effects) {
+    if (!"center" %in% names(data)) {
+      stop(paste(
+        "To include fixed center effects in the outcome model",
+        "please make sure that the input data has a 'center'",
+        "column to identify facilities."
+      ))
+    }
+    # if "center" column of the input data is not a factor,
+    # change it to factor and print a warning message
+    if (!is.factor(data$center)) {
+      data$center <- as.factor(data$center)
+      message(paste(
+        "'center' column is not a factor type. To ensure",
+        "the correct model fit, it has been converted",
+        "to the factor type."
+      ))
+    }
+  }
+
+  # preliminary check for center_weights_for_outcome_goal
+  if (!is.null(center_weights_for_outcome_goal)) {
+    # check if include center effects is set to FALSE
+    if (!include_center_effects) {
+      stop(paste(
+        "'include_center_effects' must be set to TRUE,",
+        "for values of 'center_weights_for_outcome_goal'",
+        "to be used."
+      ))
+    }
+    # check if center_weights_for_outcome_goal is a numeric vector
+    if (!is.numeric(center_weights_for_outcome_goal)) {
+      stop(paste("center_weights_for_outcome_goal must be a numeric vector."))
+    }
+    # check if center_weights_for_outcome_goal has the same length as the
+    # number of facilities provided in the input data
+    n_facilities <- length(unique(data$center))
+    if (length(center_weights_for_outcome_goal) != n_facilities) {
+      stop(paste(
+        "The length of center_weights_for_outcome_goal must match",
+        "the number of unique facilities in the input data."
+      ))
+    }
+  } else {
+    # if center_weights_for_outcome_goal is not specified, but
+    # include_center_effects is set to TRUE, we calculate its default values
+    if (include_center_effects) {
+      if (input_data_structure == "center_level") {
+        center_sizes <- tapply(
+          data$center_sample_size,
+          data$center,
+          function(x) x[1]
+        )
+        total_sample_size <- sum(center_sizes)
+      } else if (input_data_structure == "individual_level") {
+        total_sample_size <- sum(as.numeric(table(data$center)))
+        center_sizes <- as.numeric(table(data$center))
+      }
+      center_weights_for_outcome_goal <- center_sizes / total_sample_size
+    } else {
+      center_weights_for_outcome_goal <- 1
+    }
+  }
+
+  # check if values of center_weights_for_outcome_goal sum up to 1
+  if (!is.null(center_weights_for_outcome_goal)) {
+    if (sum(center_weights_for_outcome_goal) != 1) {
```

Comment:
I think for this we can consider adding a tolerance to handle floating-point precision issues such as
 if (abs(sum(center_weights_for_outcome_goal) - 1) > 1e-6) {}.
My logic: Since the numbers are in decimal forms, we can handle cases with summations being very slightly off 1. The tolerance can be higher if you want.